### ISSUE TITLE
TypeError: unsupported operand type(s) for +=: 'coo_matrix' and 'NoneType'

### ISSUE THREAD

**Issue by LachlanStuart** [metaspace2020/metaspace#509]
I've seen this exception quite a few times in the staging logs. It also affected 4 datasets out of ~3000 when Chris reprocessed all datasets on the Beta server.

```
Traceback (most recent call last):
 File "/opt/dev/metaspace/metaspace/engine/sm/engine/sm_daemons.py", line 262, in _callback
   ds=ds, annotation_job_factory=AnnotationJob, del_first=msg.get('del_first', False)
 File "/opt/dev/metaspace/metaspace/engine/sm/engine/sm_daemons.py", line 93, in annotate
   ds
 File "/opt/dev/metaspace/metaspace/engine/sm/engine/annotation_job.py", line 241, in run
   imzml_parser, self.create_mol_dbs(new_moldb_ids - completed_moldb_ids)
 File "/opt/dev/metaspace/metaspace/engine/sm/engine/annotation_job.py", line 114, in _run_annotation_jobs
   search_results_it = search_alg.search()
 File "/opt/dev/metaspace/metaspace/engine/sm/engine/msm_basic/msm_basic_search.py", line 251, in search
   results_rdd, formula_centroids.formulas_df
 File "/opt/dev/metaspace/metaspace/engine/sm/engine/msm_basic/msm_basic_search.py", line 91, in merge_results
   formula_metrics_df = pd.concat(results_rdd.map(lambda t: t[0]).collect())
 File "/usr/lib/spark/python/pyspark/rdd.py", line 814, in collect
   sock_info = self.ctx._jvm.PythonRDD.collectAndServe(self._jrdd.rdd())
 File "/opt/dev/miniconda3/envs/sm/lib/python3.6/site-packages/py4j/java_gateway.py", line 1257, in __call__
   answer, self.gateway_client, self.target_id, self.name)
 File "/opt/dev/miniconda3/envs/sm/lib/python3.6/site-packages/py4j/protocol.py", line 328, in get_return_value
   format(target_id, ".", name), value)
py4j.protocol.Py4JJavaError: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 599 in stage 1.0 failed 4 times, most recent failure: Lost task 599.3 in stage 1.0 (TID 908, 172.31.36.174, executor 1): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
 File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/worker.py", line 253, in main
   process()
 File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/worker.py", line 248, in process
   serializer.dump_stream(func(split_index, iterator), outfile)
 File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/serializers.py", line 379, in dump_stream
   vs = list(itertools.islice(iterator, batch))
 File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/util.py", line 55, in wrapper
   return f(*args, **kwargs)
 File "/opt/dev/metaspace/metaspace/engine/sm/engine/msm_basic/formula_imager.py", line 144, in process_centr_segment
   formula_images_it, compute_metrics, target_formula_inds, n_peaks, min_px
 File "/opt/dev/metaspace/metaspace/engine/sm/engine/msm_basic/formula_validator.py", line 131, in formula_image_metrics
   formula_images_buffer[f_i][p_i] += image
TypeError: unsupported operand type(s) for +=: 'coo_matrix' and 'NoneType'
```

Investigating why `image` could be `None` I also found a related issue. In https://github.com/metaspace2020/metaspace/blob/master/metaspace/engine/sm/engine/msm_basic/formula_validator.py#L134 , `p_i == n_peaks - 1` assumes that the last image isn't split across two dataset segments. If it were split, then `add_metrics` would be called twice - once with all of the data except for the last image's 2nd segment data (causing a slightly lower MSM), and a second time with only the last image's 2nd segment data (which gets discarded for not having enough images)

**Comment by intsco** [comment]
It indeed seems to be a bug. It wasn't caught in the scientific test so it must be a very seldom case.

**Comment by LachlanStuart** [comment]
This error hasn't happened since the original incident, even though several thousand datasets have been submitted since then.

I tried to reproduce both the `'coo_matrix' and 'NoneType'` bug and the potential `p_i == n_peaks - 1` issue by running the scientific test with very small dataset segments, and an assertion to ensure that `add_metrics` isn't called more than once per annotation. Some of the results changed, but it was not due to either of those bugs. 

Closing this as it's not really worth trying to fix at this point.
